packages/dd-trace/test/debugger/devtools_client/condition.spec.js

```diff
@@ -0,0 +1,261 @@
+  [
+    { contains: [{ ref: 'wset' }, { ref: 'key' }] },
+    { key: weakKey, wset: overloadMethod(new WeakSet([weakKey]), 'has') },
+    true
+  ],
+  [{ contains: [{ ref: 'map' }, 'foo'] }, { map: new Map([['foo', 'bar']]) }, true],
+  [{ contains: [{ ref: 'map' }, 'missing'] }, { map: new Map([['foo', 'bar']]) }, false],
+  [{ contains: [{ ref: 'map' }, 'foo'] }, { map: overloadMethod(new Map([['foo', 'bar']]), 'has') }, true],
+  [{ contains: [{ ref: 'wmap' }, { ref: 'key' }] }, { key: weakKey, wmap: new WeakMap([[weakKey, 'bar']]) }, true],
+  [{ contains: [{ ref: 'wmap' }, { ref: 'key' }] }, { key: {}, wmap: new WeakMap([[weakKey, 'bar']]) }, false],
+  [
+    { contains: [{ ref: 'wmap' }, { ref: 'key' }] },
+    { key: weakKey, wmap: overloadMethod(new WeakMap([[weakKey, 'bar']]), 'has') },
+    true],
+  [
+    { contains: [{ ref: 'obj' }, 'foo'] },
+    { obj: { foo: 'bar' } },
+    new TypeError('Variable obj does not support contains')
+  ],
+  [
+    { contains: [{ ref: 'obj' }, 'missing'] },
+    { obj: { foo: 'bar' } },
+    new TypeError('Variable obj does not support contains')
+  ],
+  // TODO: Ensure there's no side-effects due to Symbol.match
+  [{ matches: [{ ref: 'foo' }, '[0-9]+'] }, { foo: '42' }, true],
+  [{ matches: [{ ref: 'foo' }, { ref: 'regex' }] }, { foo: '42', regex: /[0-9]+/ }, true],
+  [{ matches: [{ ref: 'foo' }, { ref: 'regex' }] }, { foo: '42', regex: overloadMethod(/[0-9]+/, 'test') }, true],
+
+  // Undefined comparison
+  [{ eq: [{ ref: 'foo' }, { ref: 'undefined' }] }, { foo: undefined }, true],
+  [{ eq: [{ ref: 'foo' }, { ref: 'undefined' }] }, { foo: null }, false],
+  [{ eq: [{ getmember: [{ ref: 'obj' }, 'foo'] }, { ref: 'undefined' }] }, { obj: { foo: undefined } }, true],
+  [{ eq: [{ getmember: [{ ref: 'obj' }, 'foo'] }, { ref: 'undefined' }] }, { obj: {} }, true],
+  [{ eq: [{ getmember: [{ ref: 'obj' }, 'foo'] }, { ref: 'undefined' }] }, { obj: { foo: null } }, false],
+
+  // Literal values
+  [42, {}, 42],
+  [true, {}, true],
+  [{ or: [{ ref: 'bar' }, { ref: 'foo' }] }, { bar: 42 }, 42],
+  [{ and: [{ ref: 'bar' }, { ref: 'foo' }] }, { bar: 0 }, 0],
+  [{ or: [{ ref: 'bar' }, { ref: 'foo' }] }, { bar: 0 }, new ReferenceError('foo is not defined')],
+  [{ and: [{ ref: 'bar' }, { ref: 'foo' }] }, { bar: 42 }, new ReferenceError('foo is not defined')],
+  [{ isDefined: 'foo' }, { bar: 42 }, false],
+  [{ isDefined: 'bar' }, { bar: 42 }, true],
+  [{ isDefined: 'bar' }, { bar: undefined }, true],
+  // TODO: Ensure there's no side-effects due to Symbol.hasInstance
+  [{ instanceof: [{ ref: 'bar' }, 'Object'] }, { bar: {} }, true],
+  [{ instanceof: [{ ref: 'bar' }, 'Error'] }, { bar: new Error() }, true],
+  [
+    { instanceof: [{ ref: 'bar' }, 'CustomObject'] },
+    { bar: new CustomObject(), CustomObject },
+    true
+  ]
+  // TODO: Ensure there's no side-effects due to proxies
+]
+
+const definedTestCases = [
+  [{ isDefined: 'foo' }, 'const foo = undefined', false],
```

Comment:
not sure why this is false...

Reply:
The 2nd element in the array is code that's added after evaluating the the condition, so it's in in scope yet, hence it's undefined.